go/vt/concurrency/error_recorder.go

```diff
@@ -90,13 +104,53 @@ func (aer *AllErrorRecorder) RecordError(err error) {
 	aer.mu.Unlock()
 }
 
+// RecordErrorAt records a possible error at the given position
+// - does nothing if err is nil
+func (aer *AllErrorRecorder) RecordErrorAt(i int, err error) {
+	if err == nil {
+		return
+	}
+	aer.mu.Lock()
+	aer.ensureSize(i + 1)
+	aer.Errors[i] = err
+	aer.mu.Unlock()
+}
+
+// grow the internal array to the given size if needed
+func (aer *AllErrorRecorder) ensureSize(n int) {
+	if len(aer.Errors) < n {
+		errs := make([]error, n)
+		if aer.Errors != nil {
+			copy(errs, aer.Errors)
+		}
+		aer.Errors = errs
+	}
+}
+
 // HasErrors returns true if we ever recorded an error
 func (aer *AllErrorRecorder) HasErrors() bool {
 	aer.mu.Lock()
 	defer aer.mu.Unlock()
 	return len(aer.Errors) > 0
 }
 
+// GetErrors returns nil if there are no errors or
+// an array of size N with any errors.
+//
+// Note that this will grow the internal storage to the
+// given capacity if required, so any future calls to
+// RecordError will append to the array _after_ any
+// nil entries created by growing the representation.
+func (aer *AllErrorRecorder) GetErrors(n int) []error {
```

Comment:
Do we need this? I think having `multiGoTransaction` return `aer.AggrError(vterrors.Aggregate)` will do the same thing. Then we won't have to change `ExecuteMultiShard` at all.

Along the same lines, `RecordErrorAt` is currently not useful because all code paths end up using `vterrors.Aggregate` which does its own sorting. So, if you want to retain the original order, you'll need to add an option to `vterrors.Aggregate`.

Reply:
You’re right. I needed to retain the position in the original design in which the executor would need to know which shard had an error.

Now that we just append all the errors to the warning list, we can remove this and go back to the simpler collector.

Reply:
Actually just to be clear — we still need the list of errors to come out of the engine but not necessarily with the positional correspondence with the shards array.

It’s not exactly evident why in this pr but if you look at the one which puts it all together it makes more sense.

Reply:
I see. You need them for the upcoming `RecordWarnings`. Sounds good. So, we only need to get rid of `RecordErrorAt`.

Reply:
@sougou I just pushed the change to remove `RecordErrorAt`.